src/solvers/strings/string_constraint_instantiation.cpp

```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+/// \return index expressions in `expr` on `str` containing `qvar`.
+static std::unordered_set<exprt, irep_hash>
+find_indexes(const exprt &expr, const exprt &str, const symbol_exprt &qvar)
+{
+  decltype(find_indexes(expr, str, qvar)) result;
+  auto index_str_containing_qvar = [&](const exprt &e) {
+    if(auto index_expr = expr_try_dynamic_cast<index_exprt>(e))
+    {
+      const auto &arr = index_expr->array();
+      const auto str_it = std::find(arr.depth_begin(), arr.depth_end(), str);
+      return str_it != arr.depth_end() && contains(index_expr->index(), qvar);
+    }
+    return false;
+  };
+
+  std::for_each(expr.depth_begin(), expr.depth_end(), [&](const exprt &e) {
+    if(index_str_containing_qvar(e))
+      result.insert(to_index_expr(e).index());
+  });
+  return result;
+}
+
+std::map<exprt, int> map_representation_of_sum(const exprt &f)
+{
+  // number of time the leaf should be added (can be negative)
+  std::map<exprt, int> elems;
+
+  std::list<std::pair<exprt, bool>> to_process;
+  to_process.emplace_back(f, true);
+
+  while(!to_process.empty())
+  {
+    exprt cur = to_process.back().first;
+    bool positive = to_process.back().second;
+    to_process.pop_back();
+    if(cur.id() == ID_plus)
+    {
+      for(const auto &op : cur.operands())
+        to_process.emplace_back(op, positive);
+    }
+    else if(cur.id() == ID_minus)
+    {
+      to_process.emplace_back(cur.op1(), !positive);
+      to_process.emplace_back(cur.op0(), positive);
+    }
+    else if(cur.id() == ID_unary_minus)
+    {
+      to_process.emplace_back(cur.op0(), !positive);
+    }
+    else
+    {
+      if(positive)
+        elems[cur] = elems[cur] + 1;
+      else
+        elems[cur] = elems[cur] - 1;
+    }
+  }
+  return elems;
+}
+
+exprt
+sum_over_map(std::map<exprt, int> &m, const typet &type, bool negated)
+{
+  exprt sum = nil_exprt();
+  mp_integer constants = 0;
+  typet index_type;
+  if(m.empty())
+    return from_integer(0, type);
+  else
+    index_type = m.begin()->first.type();
+
+  for(const auto &term : m)
+  {
+    // We should group constants together...
+    const exprt &t = term.first;
+    int second = negated ? (-term.second) : term.second;
+    if(t.id() == ID_constant)
+    {
+      const auto int_value = numeric_cast_v<mp_integer>(to_constant_expr(t));
+      constants += int_value * second;
+    }
+    else
+    {
+      switch(second)
+      {
+        case -1:
+          if(sum.is_nil())
+            sum = unary_minus_exprt(t);
+          else
+            sum = minus_exprt(sum, t);
+          break;
+
+        case 1:
+          if(sum.is_nil())
+            sum = t;
+          else
+            sum = plus_exprt(sum, t);
+          break;
+
+        default:
+          if(second > 1)
+          {
+            if(sum.is_nil())
+              sum = t;
+            else
+              plus_exprt(sum, t);
+            for(int i = 1; i < second; i++)
+              sum = plus_exprt(sum, t);
```

Comment:
Why doesn't this use multiplication?

Reply:
I think the original intent was to have only plus_exprt and minus_exprt but this shouldn't matter for our purpose, I will replace it by multiplication to make the code simpler